File: benchmarks/crown/Input/test/test-urlparser.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "urlparser.h"

void test_parse_url() {
    URL_PARTS url_parts;

    // Test case 1: Basic HTTP URL
    assert(ParseURL("http://sullewarehouse.com/login", &url_parts));
    assert(strcmp(url_parts.scheme, "http") == 0);
    assert(strcmp(url_parts.authority, "sullewarehouse.com") == 0);
    assert(strlen(url_parts.port) == 0); // No port specified
    assert(strcmp(url_parts.path, "/login") == 0);
    assert(strlen(url_parts.query) == 0); // No query
    assert(strlen(url_parts.fragment) == 0); // No fragment

    // Test case 2: HTTPS URL with port
    assert(ParseURL("https://sullewarehouse.com:1000/login", &url_parts));
    assert(strcmp(url_parts.scheme, "https") == 0);
    assert(strcmp(url_parts.authority, "sullewarehouse.com") == 0);
    assert(strcmp(url_parts.port, "1000") == 0);
    assert(strcmp(url_parts.path, "/login") == 0);
    assert(strlen(url_parts.query) == 0);
    assert(strlen(url_parts.fragment) == 0);

    // Test case 3: URL with query parameters
    assert(ParseURL("https://sullewarehouse.com:1000/api/get?username=myuser", &url_parts));
    assert(strcmp(url_parts.scheme, "https") == 0);
    assert(strcmp(url_parts.authority, "sullewarehouse.com") == 0);
    assert(strcmp(url_parts.port, "1000") == 0);
    assert(strcmp(url_parts.path, "/api/get") == 0);
    assert(strcmp(url_parts.query, "?username=myuser") == 0);
    assert(strlen(url_parts.fragment) == 0);

    // Test case 4: URL without scheme
    assert(ParseURL("sullewarehouse.com/register", &url_parts));
    assert(strlen(url_parts.scheme) == 0); // No scheme
    assert(strcmp(url_parts.authority, "sullewarehouse.com") == 0);
    assert(strlen(url_parts.port) == 0);
    assert(strcmp(url_parts.path, "/register") == 0);
    assert(strlen(url_parts.query) == 0);
    assert(strlen(url_parts.fragment) == 0);

    // Test case 5: IPv6 URL
    assert(ParseURL("http://[2001:0db8:85a3:0000:0000:8a2e:0370:7334]/newpage", &url_parts));
    assert(strcmp(url_parts.scheme, "http") == 0);
    assert(strcmp(url_parts.authority, "2001:0db8:85a3:0000:0000:8a2e:0370:7334") == 0);
    assert(strlen(url_parts.port) == 0);
    assert(strcmp(url_parts.path, "/newpage") == 0);
    assert(strlen(url_parts.query) == 0);
    assert(strlen(url_parts.fragment) == 0);

    // Test case 6: IPv6 URL with port
    assert(ParseURL("https://[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:2678/blog", &url_parts));
    assert(strcmp(url_parts.scheme, "https") == 0);
    assert(strcmp(url_parts.authority, "2001:0db8:85a3:0000:0000:8a2e:0370:7334") == 0);
    assert(strcmp(url_parts.port, "2678") == 0);
    assert(strcmp(url_parts.path, "/blog") == 0);
    assert(strlen(url_parts.query) == 0);
    assert(strlen(url_parts.fragment) == 0);

    printf("All tests passed!\n");
}

int main() {
    test_parse_url();
    return 0;
}

File: comparisons/MinsC2Rust/crown/Output/test_project/tests/test_urlparser.rs
use test_project::urlparser::{
    ParseURL, MAX_URL_AUTHORITY_LEN, MAX_URL_FRAGMENT_LEN, MAX_URL_PATH_LEN, MAX_URL_PORT_LEN,
    MAX_URL_QUERY_LEN, MAX_URL_SCHEME_LEN, URL_PARTS,
};

use ntest::timeout;
#[test]
#[timeout(60000)]
pub fn test_parse_url() {
    let mut url_parts = URL_PARTS {
        scheme: String::new(),
        authority: String::new(),
        port: String::new(),
        path: String::new(),
        query: String::new(),
        fragment: String::new(),
    };

    // Test case 1: Basic HTTP URL
    assert!(ParseURL(
        "http://sullewarehouse.com/login".to_string(),
        &mut url_parts
    ));
    assert_eq!(url_parts.scheme, "http");
    assert_eq!(url_parts.authority, "sullewarehouse.com");
    assert_eq!(url_parts.port, ""); // No port specified
    assert_eq!(url_parts.path, "/login");
    assert_eq!(url_parts.query, ""); // No query
    assert_eq!(url_parts.fragment, ""); // No fragment

    // Test case 2: HTTPS URL with port
    assert!(ParseURL(
        "https://sullewarehouse.com:1000/login".to_string(),
        &mut url_parts
    ));
    assert_eq!(url_parts.scheme, "https");
    assert_eq!(url_parts.authority, "sullewarehouse.com");
    assert_eq!(url_parts.port, "1000");
    assert_eq!(url_parts.path, "/login");
    assert_eq!(url_parts.query, "");
    assert_eq!(url_parts.fragment, "");

    // Test case 3: URL with query parameters
    assert!(ParseURL(
        "https://sullewarehouse.com:1000/api/get?username=myuser".to_string(),
        &mut url_parts
    ));
    assert_eq!(url_parts.scheme, "https");
    assert_eq!(url_parts.authority, "sullewarehouse.com");
    assert_eq!(url_parts.port, "1000");
    assert_eq!(url_parts.path, "/api/get");
    assert_eq!(url_parts.query, "?username=myuser");
    assert_eq!(url_parts.fragment, "");

    // Test case 4: URL without scheme
    assert!(ParseURL(
        "sullewarehouse.com/register".to_string(),
        &mut url_parts
    ));
    assert_eq!(url_parts.scheme, ""); // No scheme
    assert_eq!(url_parts.authority, "sullewarehouse.com");
    assert_eq!(url_parts.port, "");
    assert_eq!(url_parts.path, "/register");
    assert_eq!(url_parts.query, "");
    assert_eq!(url_parts.fragment, "");

    // Test case 5: IPv6 URL
    assert!(ParseURL(
        "http://[2001:0db8:85a3:0000:0000:8a2e:0370:7334]/newpage".to_string(),
        &mut url_parts
    ));
    assert_eq!(url_parts.scheme, "http");
    assert_eq!(
        url_parts.authority,
        "2001:0db8:85a3:0000:0000:8a2e:0370:7334"
    );
    assert_eq!(url_parts.port, "");
    assert_eq!(url_parts.path, "/newpage");
    assert_eq!(url_parts.query, "");
    assert_eq!(url_parts.fragment, "");

    // Test case 6: IPv6 URL with port
    assert!(ParseURL(
        "https://[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:2678/blog".to_string(),
        &mut url_parts
    ));
    assert_eq!(url_parts.scheme, "https");
    assert_eq!(
        url_parts.authority,
        "2001:0db8:85a3:0000:0000:8a2e:0370:7334"
    );
    assert_eq!(url_parts.port, "2678");
    assert_eq!(url_parts.path, "/blog");
    assert_eq!(url_parts.query, "");
    assert_eq!(url_parts.fragment, "");

    println!("All tests passed!");
}
